### DISCUSSION TITLE
Where is `PendingDeprecationWarning` declared in the SQLA source?

### DISCUSSION THREAD

**Discussion by jenstroeger** [sqlalchemy/sqlalchemy#10400] [Usage Questions] [answered]
Hello,

I’m reading through SQLA’s [exc.py](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/exc.py) module and I noticed:

https://github.com/sqlalchemy/sqlalchemy/blob/93146598a49d4a56e0bc0e0232ea9a9467250543/lib/sqlalchemy/exc.py#L820-L824

Where is the base class `PendingDeprecationWarning` defined? I can’t find it declared in the code
```
sqlalchemy > rg PendingDeprecationWarning
lib/sqlalchemy/exc.py
820:class SAPendingDeprecationWarning(PendingDeprecationWarning):

lib/sqlalchemy/testing/warnings.py
32:    PendingDeprecationWarning have no effect if added here, since pytest
34:    ``always::PendingDeprecationWarning`` and ``always::DeprecationWarning``

doc/build/changelog/changelog_05.rst
2378:      Fixed PendingDeprecationWarning involving order_by parameter

test/base/test_except.py
513:    ([sa_exceptions.SAPendingDeprecationWarning], [lambda cls: cls(1, 2, 3)]),
```
but it must be somewhere, otherwise I’d expect a [NameError](https://docs.python.org/3/library/exceptions.html#NameError) to be raised by Python upon loading SQLA.

What am I missing?

**Comment by CaselIT** [comment, marked as answer]
Hi,

It's a python builtin warning

**Reply by jenstroeger** [reply]
So it is: https://docs.python.org/3/library/exceptions.html#PendingDeprecationWarning 🤦🏻‍♂️

**Reply by CaselIT** [reply]
it's not very popular since iirc it's ignored by pretty much anything
